addons/web_hierarchy/static/src/hierarchy_model.js

```diff
@@ -920,156 +949,184 @@ export class HierarchyModel extends Model {
-                }
-            }
-            node.hidden = true;
-            this.notify({ scrollTarget: "none" });
-            await this.mutex.exec(async () => {
-                try {
-                    await this.updateParentId(node, parentResId || parentNode?.resId);
-                    node.hidden = false;
-                    if (parentNode) {
-                        if (node.id === node.tree.root.id) {
-                            this.root.removeTree(node.tree);
-                            node.tree = parentNode.tree;
-                        } else {
-                            node.removeParentNode();
-                        }
-                    } else {
-                        node.removeParentNode();
-                        this.root.addNewRootNode(node);
-                    }
-                } catch (error) {
-                    node.hidden = false;
-                    this.notify({ scrollTarget: "none" });
-                    if (error.exceptionName === "odoo.exceptions.ValidationError") {
-                        this.notification.add(
-                            error.data.message,
-                            { type: "danger" },
-                        )
-                        return;
-                    }
-                    throw error;
-                }
-            });
-            if (descendantsParentIds.length) {
-                domain = Domain.or([domain, [[this.parentFieldName, "in", descendantsParentIds]]]);
+                throw error;
             }
-            domain = domain.toList({});
-            const descendants = {};
-            if (domain.length) {
-                const data = await this.orm.searchRead(
-                    this.resModel,
-                    domain,
-                    this.fieldsToFetch,
-                    { context: this.config.context },
+        });
+        // Reload impacted records.
+        const domain = this.computeUpdateParentNodeDomain(node, parentResId, parentNode);
+        const data = await this.orm.searchRead(this.resModel, domain, this.fieldsToFetch, {
+            context: this.config.context,
+            order: orderByToString(this.config.orderBy),
+        });
+        const formattedData = this._formatData(data);
+        // Validate that data coming from the server is still compatible with the current
+        // configuration of the hierarchy.
+        for (const record of formattedData) {
+            if (getIdOfMany2oneField(record[this.parentFieldName]) !== parentResId) {
+                node.hidden = false;
+                this.notify({ scrollTarget: "none" });
+                this.notification.add(
+                    _t(
+                        `The parent of "%s" was successfully updated. Reloading records to account for other changes.`,
+                        node.displayName
+                    ),
+                    { type: "success" }
```

Comment:
I am wondering if it should not be acceptable to not reload the records each time we update the parent.

I don't think it is really needed to be sure the order is still respected when we update the parent. 

Moreover, as you say in your commit message, it is very costly to keep the right order each time we update the parent. 😕 

What do you think? 🙂

Reply:
As discussed privately, not guaranteeing the order means that there could be weird re-ordering behavior, i.e.:
dropping a card in knowledge, clicking on "favorite" button => the view reloads and the records are re-ordered.

The current code will be kept for now even if perf expensive, until a rework is done that will use reactivity (like Record for kanban/list/...).